metal/eip_controlplane_reconciliation.go

```diff
@@ -267,163 +279,349 @@ func (m *controlPlaneEndpointManager) reassign(ctx context.Context, nodes []*v1.
 	}
+
 	if len(controlPlaneEndpoint.Assignments) > 1 {
-		return fmt.Errorf("the elastic ip %s has more than one node assigned to it and this is currently not supported. Fix it manually unassigning devices", controlPlaneEndpoint.ID)
+		return nil, fmt.Errorf("the elastic ip %s has more than one node assigned to it and this is currently not supported. Fix it manually unassigning devices", controlPlaneEndpoint.ID)
 	}
 
-	// for ease of use
-	eip := controlPlaneEndpoint.Address
+	return controlPlaneEndpoint, nil
+}
 
-	for _, svc := range svcs {
-		// only take default/kubernetes
-		if svc.Namespace != "default" || svc.Name != "kubernetes" {
-			continue
+func (m *controlPlaneEndpointManager) nodeIsAssigned(ctx context.Context, node *v1.Node, ipReservation *packngo.IPAddressReservation) (bool, error) {
+	for _, a := range ipReservation.Assignments {
+		for _, na := range node.Status.Addresses {
+			if na.Address == a.Address {
+				return true, nil
+			}
 		}
+	}
 
-		// get the target port
-		existingPorts := svc.Spec.Ports
-		if len(existingPorts) < 1 {
-			return errors.New("default/kubernetes service does not have any ports defined")
+	return false, nil
+}
+
+func tryFilterSelf(self *v1.Node, nodes []*v1.Node) []*v1.Node {
+	var remainingNodes []*v1.Node
+
+	for i := range nodes {
+		if nodes[i].Name != self.Name {
+			remainingNodes = append(remainingNodes, nodes[i])
 		}
+	}
 
-		// track which port the kube-apiserver actually is listening on
-		m.nodeAPIServerPort = existingPorts[0].TargetPort.IntVal
-		// did we set a specific port, or did we request that it just be left as is?
-		if m.apiServerPort == 0 {
-			m.apiServerPort = m.nodeAPIServerPort
+	if len(remainingNodes) > 0 {
+		return remainingNodes
+	}
+
+	return nodes
+}
+
+func filterDeletingNodes(nodes []*v1.Node) []*v1.Node {
+	var liveNodes []*v1.Node
+	for i := range nodes {
+		if nodes[i].DeletionTimestamp.IsZero() {
+			liveNodes = append(liveNodes, nodes[i])
 		}
+	}
 
-		// get the endpoints for this service
-		eps := m.k8sclient.CoreV1().Endpoints(svc.Namespace)
-		ep, err := eps.Get(ctx, svc.Name, metav1.GetOptions{})
-		if err != nil {
-			klog.V(2).Infof("failed to get endpoints %s: %v", svc.Name, err)
-			return fmt.Errorf("failed to get endpoints %s: %v", svc.Name, err)
+	return liveNodes
+}
+
+func tryFilterUnschedulableNodes(nodes []*v1.Node) []*v1.Node {
+	var schedulableNodes []*v1.Node
+	for i := range nodes {
+		if nodes[i].Spec.Unschedulable {
+			continue
 		}
-		// two options:
-		// - our endpoints already exists: just copy the endpoints
-		// - our endpoints does not exist: create it
-		epExisted := true
-		myeps := m.k8sclient.CoreV1().Endpoints(externalServiceNamespace)
-		myep, err := myeps.Get(ctx, externalServiceName, metav1.GetOptions{})
+
+		schedulableNodes = append(schedulableNodes, nodes[i])
+	}
+
+	if len(schedulableNodes) > 0 {
+		return schedulableNodes
+	}
+
+	return nodes
+}
+
+type nodeFitler func([]*v1.Node) []*v1.Node
```

Comment:
I think you meant for this to be `nodeFilter` (instead of `nodeFitler`, `t` and `l` reversed). Same for where it is consumed in a two other places

Reply:
😬 oops, yeah. I have no idea what a fitler is...